Add SQS Explorer Node

ranyoo2367 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Added SQS to the AWS Explorer. 


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
<!--- Describe your changes in detail -->
The SQS root node is added to the AWS Explorer. SQS root node lists the names of all available queues in the selected region as child nodes. There are no icons added yet, but will be added in the near future. The user is able to right-click and copy ARN.

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Functional unit tests were conducted to simulate an SQS client with a few queues, no queues, and error loading queues. A small test to return queue name and ARN was added to ensure URL parsing was successful.

## Screenshots (if appropriate)
![SqsExplorerNodes](https://user-images.githubusercontent.com/30319724/87190745-0f561d80-c2c1-11ea-9698-52216a16f9fc.gif)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/SqsExplorerNodes.kt:
```diff
@@ -0,0 +1,40 @@
+    project,
+    SqsClient.SERVICE_NAME,
+    queueUrl,
+    EmptyIcon.ICON_0 // TODO: Get & change icons
```
nit: `AllIcons.Nodes.EmptyNode` is 16x16 and empty

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/SqsServiceNodeTest.kt:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Test
+    fun noQueuesListed() {
```
nit: all of our new test have been \`fluent test name`

For example:
```kt
@Test
fun `No queues listed`() {
...
}
```

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/SqsWindow.kt:
```diff
@@ -0,0 +1,4 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.sqs
```
Accidental inclusion?

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/Queue.kt:
```diff
@@ -0,0 +1,11 @@
+
+class Queue(queueUrl: String) {
+    val regionId = queueUrl.substringAfter("sqs.").substringBefore('.')
+    val accountId = queueUrl.substringAfter("amazonaws.com/").substringBefore('/')
```
China is `sqs.<region>.amazonaws.com.cn`: https://docs.amazonaws.cn/en_us/general/latest/gr/cnnorth_region.html
and govcloud is also different, so we can't rely on amazonaws.com. This doc has the format: https://docs.amazonaws.cn/sdk-for-net/v3/developer-guide/how-to/sqs/QueueURL.html

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/Queue.kt:
```diff
@@ -0,0 +1,11 @@
+class Queue(queueUrl: String) {
+    val regionId = queueUrl.substringAfter("sqs.").substringBefore('.')
+    val accountId = queueUrl.substringAfter("amazonaws.com/").substringBefore('/')
+    val queueName = queueUrl.substringAfter("$accountId/")
```
We should throw an exception if any of these are empty (or fail in some way). It would also be good to add tests for the functions along with giving them China and Govcloud urls

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/SqsExplorerNodes.kt:
```diff
@@ -0,0 +1,40 @@
+import software.aws.toolkits.jetbrains.services.sqs.resources.SqsResources
+
+class SqsServiceNode(project: Project, service: AwsExplorerServiceNode) :
+    CacheBackedAwsExplorerServiceRootNode<String>(project, service, SqsResources.LIST_QUEUE_URLS) {
```
spacing nit: would prefer line return after ( and have project/service on their own lines

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/SqsServiceNodeTest.kt:
```diff
@@ -0,0 +1,80 @@
+        assertThat(children).allMatch { it is AwsExplorerErrorNode }
+    }
+
+    private fun resourceCache() = MockResourceCache.getInstance(projectRule.project)
```
We now have `MockResourceCacheRule`, so use that. It will take care of cleanup so you don't need the `@Before`

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/SqsServiceNodeTest.kt:
```diff
@@ -0,0 +1,80 @@
+    @Test
+    fun sqsQueuesAreListed() {
+        val queueList = listOf(
+            "https://sqs.us-east-1.amazonaws.com/123456789012/test2",
```
Nit: could probably just do 2 queues, but the size does not matter either way

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/Queue.kt:
```diff
@@ -0,0 +1,11 @@
+    val regionId = queueUrl.substringAfter("sqs.").substringBefore('.')
+    val accountId = queueUrl.substringAfter("amazonaws.com/").substringBefore('/')
+    val queueName = queueUrl.substringAfter("$accountId/")
+    val arn = "arn:aws:sqs:$regionId:$accountId:$queueName"
```
This needs to be partition award (can't use `aws`)

ranyoo2367 commented:
- Modified the Queue.kt to accommodate different ARN and URL formats for China and GovCloud (US) regions. Added exceptions for handling parsing errors.
- Added QueueTest.kt to test queue URL parsing for China, GovCloud, and remaining regions as they handle different partition IDs. Error handling is tested as well.
- Added MockResourceCacheRule for SqsServiceNodeTest.kt
- nit changes

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/SqsServiceNodeTest.kt:
```diff
@@ -5,40 +5,41 @@ package software.aws.toolkits.jetbrains.services.sqs
     @Test
-    fun sqsQueuesAreListed() {
+    fun `Sqs queues are listed`() {
         val queueList = listOf(
```
nit: since this is never reused, why give it a name? Could just put it into the resource cache call

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/QueueTest.kt:
```diff
@@ -0,0 +1,41 @@
+        val queueRegion = AwsRegion("cn-northwest-1", "China (Ningxia)", "aws-cn")
+        val queue = Queue("https://sqs.cn-northwest-1.amazonaws.com.cn/123456789012/test-1", queueRegion)
+        val queueArn = "arn:aws-cn:sqs:cn-northwest-1:123456789012:test-1"
+        assertThat(queue.arn).isEqualTo(queueArn)
```
Should also add asserts for name and account id

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/QueueTest.kt:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Test
+    fun `Common region endpoint parsed`() {
```
nit: Common -> AWS or Public

hunterwerlla commented on jetbrains-core/tst/software/aws/toolkits/jetbrains/services/sqs/QueueTest.kt:
```diff
@@ -0,0 +1,41 @@
+    @Test
+    fun `Throw exception with invalid url`() {
+        val queueRegion = AwsRegion("us-east-1", "US East (N. Virginia)", "aws")
+        assertFailsWith<IllegalArgumentException> { val queue = Queue("Not a URL", queueRegion) }
```
Should also test throwing exceptions when it is a url but the constituent parts have issues

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/Queue.kt:
```diff
@@ -3,9 +3,24 @@
+    val queueName = queueUrl.substringAfter("$accountId/") ?: throw IllegalArgumentException()
+    val arn = "arn:${region.partitionId}:sqs:${region.id}:$accountId:$queueName"
+
+    fun String.substringAfter(delimiter: String): String? {
```
After seeing this implementation in action, would prefer something like this so that it can have multiple lines of logic vs using the ?:

```kt
    var jetbrainsChangeNotesFile: File?
        get() {
            ...
            ...
            return thing
        }
```

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/Queue.kt:
```diff
@@ -3,9 +3,24 @@
+    val accountId = queueUrl.substringAfter("${region.id}")
+        ?.substringAfter("/")
+        ?.substringBefore("/")
+        ?: throw IllegalArgumentException()
```
Should add a message for the exception

ranyoo2367 commented:
- Revised Queue.kt to use get() and make logic more clear
- Added asserts for account ID and queue name for QueueTest.kt. Added new tests for invalid account ID or queue name in the URL
- Added message for exception

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/Queue.kt:
```diff
@@ -0,0 +1,31 @@
+import java.lang.IllegalArgumentException
+import software.aws.toolkits.resources.message
+
+/* This does not support Federal Information Processing Standard (FIPS) */
```
nit: could just put FIPS, don't need to spell out the acronym

hunterwerlla commented on jetbrains-core/src/software/aws/toolkits/jetbrains/services/sqs/Queue.kt:
```diff
@@ -0,0 +1,31 @@
+
+/* This does not support Federal Information Processing Standard (FIPS) */
+class Queue(val queueUrl: String, val region: AwsRegion) {
+    val accountId: String = queueUrl.substringAfter("${region.id}").substringAfter("/").substringBefore("/")
```
Hm, would prefer `by lazy`
